src/main/java/eu/europa/ted/eforms/xpath/XPathListenerImpl.java

```diff
@@ -0,0 +1,168 @@
+    final ParseTreeWalker walker = new ParseTreeWalker();
+    walker.walk(this, tree);
+
+    steps.stream().forEach(s -> {
+      XPathStep step = new XPathStep(s.stepText, s.predicates);
+      xpathInfo.addStep(step);
+    });
+
+    if (!xpathInfo.isAttribute()) {
+      // The XPath does not point to an attribute, so it is the path to the last element
+      xpathInfo.setPathToLastElement(xpathInput);
+    }
+
+    return xpathInfo;
+  }
+
+  @Override
+  public void exitAxisstep(AxisstepContext ctx) {
+    if (inPredicateMode()) {
+      return;
+    }
+
+    // When we recognize a step, we add it to the queue if is is empty.
+    // If the queue is not empty, and the depth of the new step is not smaller than
+    // the depth of the last step in the queue, then this step needs to be added to
+    // the queue too.
+    // Otherwise, the last step in the queue is a sub-expression of the new step,
+    // and we need to
+    // replace it in the queue with the new step.
+    if (this.steps.isEmpty() || !this.steps.getLast().isPartOf(ctx.getSourceInterval())) {
+      this.steps.offer(new StepInfo(ctx, this::getInputText));
+    } else {
+      Interval removedInterval = ctx.getSourceInterval();
+      while(!this.steps.isEmpty() && this.steps.getLast().isPartOf(removedInterval)) {
+        this.steps.removeLast();
+      }
+      this.steps.offer(new StepInfo(ctx, this::getInputText));
+    }
+  }
+
+  @Override
+  public void exitFilterexpr(FilterexprContext ctx) {
+    if (inPredicateMode()) {
+      return;
+    }
+
+    // Same logic as for axis steps here (sse exitAxisstep).
+    if (this.steps.isEmpty() || !this.steps.getLast().isPartOf(ctx.getSourceInterval())) {
+      this.steps.offer(new StepInfo(ctx, this::getInputText));
+    } else {
+      Interval removedInterval = ctx.getSourceInterval();
+      while(!this.steps.isEmpty() && this.steps.getLast().isPartOf(removedInterval)) {
+        this.steps.removeLast();
+      }
+      this.steps.offer(new StepInfo(ctx, this::getInputText));
+    }
+  }
+
+  @Override
+  public void enterPredicate(PredicateContext ctx) {
+    this.inPredicate++;
+  }
+
+  @Override
+  public void exitPredicate(PredicateContext ctx) {
+    this.inPredicate--;
+  }
+
+  @Override
+  public void exitAbbrevforwardstep(AbbrevforwardstepContext ctx) {
+    if (!inPredicateMode() && ctx.AT() != null) {
+      xpathInfo.setAttributeName(ctx.nodetest().getText());
+
+      int splitPosition = ctx.AT().getSymbol().getCharPositionInLine();
+      String path = inputText.substring(0, splitPosition);
+      while (path.endsWith("/")) {
+        path = path.substring(0, path.length() - 1);
+      }
+      xpathInfo.setPathToLastElement(path);
+    }
+  }
+
+  /**
+   * Helper method that returns the input text that matched a parser rule context. It is useful
+   * because {@link ParserRuleContext#getText()} omits whitespace and other lexer tokens in the
+   * HIDDEN channel.
+   *
+   * @param context
+   * @return
```

Comment:
This Java docs has empty param and return.

Reply:
This was copied from the code in EFX Toolkit.
Fixed in added commit.